Sources/Common/DataModel/Polygon/index.js

```diff
@@ -334,122 +1127,162 @@ function vtkPolygon(publicAPI, model) {
+
   publicAPI.triangulate = () => {
-    if (!model.firstPoint) {
+    if (model.points.getNumberOfPoints() === 0) {
       return null;
     }
 
     return earCutTriangulation();
   };
 
-  publicAPI.setPoints = (points) => {
-    model.pointCount = points.length;
-
-    model.firstPoint = {
-      id: 0,
-      point: points[0],
-      next: null,
-      previous: null,
-    };
-
-    let currentPoint = model.firstPoint;
-    for (let i = 1; i < model.pointCount; i++) {
-      currentPoint.next = {
-        id: i,
-        point: points[i],
-        next: null,
-        previous: currentPoint,
-      };
-      currentPoint = currentPoint.next;
-    }
-
-    model.firstPoint.previous = currentPoint;
-    currentPoint.next = model.firstPoint;
+  publicAPI.setPoints = (points, pointsIds) => {
+    let pointsData = points;
+    if (Array.isArray(pointsData[0])) {
+      pointsData = points.flat(2);
+    }
+    model.points = vtkPoints.newInstance({ values: pointsData });
+    if (!pointsIds) {
+      model.pointsIds = new Array(pointsData.length / 3);
+      for (let i = 0; i < points.length; i++) model.pointsIds[i] = i;
+    } else {
+      model.pointsIds = pointsIds;
+    }
+    publicAPI.computeNormal();
   };
 
-  publicAPI.getPointArray = () => model.tris;
+  publicAPI.computeCentroid = () => computeCentroid(model.points, model.normal);
+
+  //------------------------------------------------------------------------------
+  // Compute the area of the polygon (oriented in 3D space). It uses an
+  // efficient approach where the area is computed in 2D and then projected into
+  // 3D space.
+  publicAPI.computeArea = () => computeArea(model.points, model.normal);
+
+  publicAPI.distanceToPolygon = (x, closest) =>
+    distanceToPolygon(x, model.points, closest);
+
+  publicAPI.isConvex = () => isConvex(model.points);
+
+  publicAPI.pointInPolygon = (point) =>
+    pointInPolygon(
+      point,
+      model.points.getData(),
+      publicAPI.getBounds(),
+      model.normal
+    );
+
+  //------------------------------------------------------------------------------
+  // Compute interpolation weights using 1/r**2 normalized sum or mean value
+  // coordinate.
+  publicAPI.interpolateFunctions = (point, useMVCInterpolation) =>
+    interpolateFunctions(point, model.points, useMVCInterpolation);
+
+  //------------------------------------------------------------------------------
+  //
+  // Intersect this polygon with finite line defined by p1 & p2
+  //
+  publicAPI.intersectWithLine = (x1, x2) =>
+    intersectWithLine(x1, x2, model.points, model.normal);
+
+  publicAPI.intersectConvex2DCells = (cell) =>
+    intersectConvex2DCells(cell, model.points, model.normal);
 }
 
 // ----------------------------------------------------------------------------
 // Object factory
 // ----------------------------------------------------------------------------
 
 const DEFAULT_VALUES = {
-  firstPoint: null,
-  pointCount: 0,
-  tris: [],
+  points: [],
```

Comment:
model.points is actually defined in vtkCell as vtkPoints. I doubt it should be [] here